frame/staking-rewards/src/lib.rs

```diff
@@ -494,11 +515,77 @@ pub mod pallet {
 
 		#[transactional]
 		fn unstake(
-			_who: &Self::AccountId,
-			_position: &Self::PositionId,
+			who: &Self::AccountId,
+			position_id: &Self::PositionId,
 			_remove_amount: Self::Balance,
 		) -> DispatchResult {
-			Err("Not implemented".into())
+			let keep_alive = false;
+			let stake = Stakes::<T>::try_get(position_id).map_err(|_| Error::<T>::StakeNotFound)?;
+			ensure!(who == &stake.owner, Error::<T>::OnlyStakeOwnerCanUnstake);
+			ensure!(!stake.reductions.is_empty(), Error::<T>::ReductionConfigProblem);
+			let early_unlock = stake.lock.started_at.safe_add(&stake.lock.duration)? >
+				T::UnixTime::now().as_secs();
+			let pool_id = stake.reward_pool_id;
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let mut inner_rewards = rewards_pool.rewards.into_inner();
+			for (asset_id, reward) in inner_rewards.iter_mut() {
+				let inflation = match stake.reductions.get(asset_id) {
+					Some(inflation) => *inflation,
+					None => Zero::zero(),
+				};
+				let claim = if rewards_pool.total_shares == Zero::zero() {
+					Zero::zero()
+				} else {
+					reward
+						.total_rewards
+						.safe_mul(&stake.share)?
+						.safe_div(&rewards_pool.total_shares)?
+						.safe_sub(&inflation)?
+				};
+
+				let claim = sp_std::cmp::min(
+					(Perbill::one() - stake.lock.unlock_penalty).mul_ceil(claim),
+					reward.total_rewards.safe_sub(&reward.claimed_rewards)?,
+				);
+
+				reward.claimed_rewards = reward.claimed_rewards.safe_add(&claim)?;
+
+				T::Assets::transfer(
+					rewards_pool.asset_id,
```

Comment:
```suggestion
					reward.asset_id,
```

Reply:
Claim is for rewarded asset not for the staked asset

Reply:
Fixed.